File: agent/core/fixture_scanner.py
# agent/core/fixture_scanner.py

"""Project test-fixture / helper symbol scanner.

Scans the user's test fixture and helper modules (e.g. tests/fixtures/,
test-utils/, helpers/) and extracts the named exports from each file.
This list is injected into the generation prompt as a "Project Symbols"
section so the LLM imports real identifiers from the project instead of
inventing plausible-sounding ones (#62).

Lightweight by design: regex-based, no external parser dependency. The
goal is to ground the LLM, not to produce a perfect AST.
"""

import re
from dataclasses import dataclass, field
from pathlib import Path
from typing import Dict, List


# Where projects typically keep test helpers and fixtures. Order matters
# only insofar as we deduplicate by resolved path.
_FIXTURE_DIRS = (
    "tests/fixtures",
    "tests/test-utils",
    "tests/helpers",
    "tests/support",
    "test/fixtures",
    "test/test-utils",
    "test/helpers",
    "e2e/fixtures",
    "e2e/helpers",
    "__tests__/fixtures",
    "__tests__/helpers",
    "fixtures",
    "test-utils",
)

_IGNORED_DIRS = {
    "node_modules", ".git", "__pycache__", ".venv", "venv",
    "dist", "build", ".next", ".nuxt", "coverage",
}

_MAX_FILES = 20
_MAX_FILE_BYTES = 32_768
_MAX_SYMBOLS_PER_FILE = 12

# TS/JS: `export const FOO`, `export function bar`, `export class Baz`,
# `export let x`, `export async function y`, `export default function z`.
_TS_DECL_RE = re.compile(
    r"^\s*export\s+(?:default\s+)?"
    r"(?:async\s+)?"
    r"(?:const|let|var|function|class|interface|type|enum)\s+"
    r"([A-Za-z_$][\w$]*)",
    re.MULTILINE,
)

# TS/JS: `export { foo, bar as baz }`.
_TS_NAMED_RE = re.compile(r"^\s*export\s*\{([^}]+)\}", re.MULTILINE)

# Python: top-level `def foo` / `class Bar` (skip dunder/private).
_PY_DECL_RE = re.compile(
    r"^(?:def|class)\s+([A-Za-z][\w]*)",
    re.MULTILINE,
)


@dataclass
class FixtureSymbols:
    """Map of import-style path -> ordered list of exported symbols."""

    # Relative path (POSIX, project-rooted) -> list of exported names.
    by_module: Dict[str, List[str]] = field(default_factory=dict)
    files_scanned: int = 0

    @property
    def is_empty(self) -> bool:
        return not self.by_module


class FixtureScanner:
    """Extracts exported symbols from project test-fixture/helper modules."""

    def scan(self, project_root: str = ".") -> FixtureSymbols:
        root = Path(project_root).resolve()
        result = FixtureSymbols()
        if not root.exists():
            return result

        candidates: List[Path] = []
        seen: set = set()
        for rel in _FIXTURE_DIRS:
            d = root / rel
            if not d.is_dir():
                continue
            if d in seen:
                continue
            seen.add(d)
            for path in sorted(d.rglob("*")):
                if not path.is_file():
                    continue
                if any(part in _IGNORED_DIRS for part in path.parts):
                    continue
                if path.suffix not in (".ts", ".tsx", ".js", ".jsx", ".py"):
                    continue
                candidates.append(path)
                if len(candidates) >= _MAX_FILES:
                    break
            if len(candidates) >= _MAX_FILES:
                break

        for path in candidates:
            try:
                text = path.read_text(encoding="utf-8", errors="ignore")
            except OSError:
                continue
            if len(text) > _MAX_FILE_BYTES:
                text = text[:_MAX_FILE_BYTES]

            symbols = (
                _extract_python(text) if path.suffix == ".py"
                else _extract_ts(text)
            )
            if not symbols:
                continue

            rel_path = path.relative_to(root).as_posix()
            result.by_module[rel_path] = symbols[:_MAX_SYMBOLS_PER_FILE]
            result.files_scanned += 1

        return result


def _extract_ts(text: str) -> List[str]:
    """Return ordered list of exported names from TS/JS source text."""
    seen: List[str] = []
    seen_set: set = set()

    def add(name: str) -> None:
        name = name.strip()
        if not name or name in seen_set:
            return
        seen_set.add(name)
        seen.append(name)

    for m in _TS_DECL_RE.finditer(text):
        add(m.group(1))

    for m in _TS_NAMED_RE.finditer(text):
        # "foo, bar as baz, qux" -> ["bar" (the rebound name wins via `as`)]
        for piece in m.group(1).split(","):
            piece = piece.strip()
            if not piece:
                continue
            if " as " in piece:
                _, _, alias = piece.partition(" as ")
                add(alias)
            else:
                add(piece)

    return seen


def _extract_python(text: str) -> List[str]:
    """Return ordered list of top-level public names from Python source."""
    seen: List[str] = []
    seen_set: set = set()
    for m in _PY_DECL_RE.finditer(text):
        name = m.group(1)
        if name.startswith("_"):
            continue
        if name in seen_set:
            continue
        seen_set.add(name)
        seen.append(name)
    return seen

File: ts/src/core/fixture-scanner.ts
/**
 * Project test-fixture / helper symbol scanner — faithful TS port of
 * `agent/core/fixture_scanner.py`.
 *
 * Extracts named exports from files under conventional fixture/helper dirs so
 * downstream generation imports real identifiers. Regex-based, no AST.
 */

import { existsSync, readdirSync, readFileSync, statSync } from 'node:fs';
import { relative, resolve } from 'node:path';

const FIXTURE_DIRS = [
  'tests/fixtures',
  'tests/test-utils',
  'tests/helpers',
  'tests/support',
  'test/fixtures',
  'test/test-utils',
  'test/helpers',
  'e2e/fixtures',
  'e2e/helpers',
  '__tests__/fixtures',
  '__tests__/helpers',
  'fixtures',
  'test-utils',
];

const IGNORED_DIRS = new Set([
  'node_modules',
  '.git',
  '__pycache__',
  '.venv',
  'venv',
  'dist',
  'build',
  '.next',
  '.nuxt',
  'coverage',
]);

const SUPPORTED_EXTS = ['.ts', '.tsx', '.js', '.jsx', '.py'];
const MAX_FILES = 20;
const MAX_FILE_BYTES = 32_768;
const MAX_SYMBOLS_PER_FILE = 12;

const TS_DECL_RE =
  /^\s*export\s+(?:default\s+)?(?:async\s+)?(?:const|let|var|function|class|interface|type|enum)\s+([A-Za-z_$][\w$]*)/gm;
const TS_NAMED_RE = /^\s*export\s*\{([^}]+)\}/gm;
const PY_DECL_RE = /^(?:def|class)\s+([A-Za-z][\w]*)/gm;

export interface FixtureSymbols {
  byModule: Record<string, string[]>;
  filesScanned: number;
}

export function isEmpty(s: FixtureSymbols): boolean {
  return Object.keys(s.byModule).length === 0;
}

export class FixtureScanner {
  scan(projectRoot = '.'): FixtureSymbols {
    const root = resolve(projectRoot);
    const result: FixtureSymbols = { byModule: {}, filesScanned: 0 };
    if (!existsSync(root)) return result;

    const candidates = this.collectCandidates(root);

    for (const path of candidates) {
      let text = safeRead(path);
      if (text === null) continue;
      if (text.length > MAX_FILE_BYTES) text = text.slice(0, MAX_FILE_BYTES);

      const symbols = path.endsWith('.py')
        ? extractPython(text)
        : extractTs(text);
      if (symbols.length === 0) continue;

      const relPath = relative(root, path).split(/[/\\]/).join('/');
      result.byModule[relPath] = symbols.slice(0, MAX_SYMBOLS_PER_FILE);
      result.filesScanned += 1;
    }

    return result;
  }

  private collectCandidates(root: string): string[] {
    const candidates: string[] = [];
    const seen = new Set<string>();
    for (const rel of FIXTURE_DIRS) {
      const fixtureDir = resolve(root, rel);
      if (!isDir(fixtureDir) || seen.has(fixtureDir)) continue;
      seen.add(fixtureDir);
      for (const path of rglobFilesSorted(fixtureDir)) {
        if (path.split(/[/\\]/).some((part) => IGNORED_DIRS.has(part)))
          continue;
        if (!SUPPORTED_EXTS.some((e) => path.endsWith(e))) continue;
        candidates.push(path);
        if (candidates.length >= MAX_FILES) return candidates;
      }
      if (candidates.length >= MAX_FILES) break;
    }
    return candidates;
  }
}

/** Recursive file list, sorted — mirrors Python's `sorted(dir.rglob("*"))`. */
function rglobFilesSorted(dir: string): string[] {
  const files: string[] = [];
  const walk = (d: string): void => {
    let entries;
    try {
      entries = readdirSync(d, { withFileTypes: true });
    } catch {
      return;
    }
    for (const entry of entries) {
      const full = resolve(d, entry.name);
      if (entry.isDirectory()) walk(full);
      else if (entry.isFile()) files.push(full);
    }
  };
  walk(dir);
  return files.sort(cmp);
}

function extractTs(text: string): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  const add = (nameRaw: string): void => {
    const name = nameRaw.trim();
    if (!name || seen.has(name)) return;
    seen.add(name);
    out.push(name);
  };

  for (const m of text.matchAll(TS_DECL_RE)) add(m[1]!);

  for (const m of text.matchAll(TS_NAMED_RE)) {
    for (const pieceRaw of m[1]!.split(',')) {
      const piece = pieceRaw.trim();
      if (!piece) continue;
      const asIdx = piece.indexOf(' as ');
      if (asIdx !== -1) add(piece.slice(asIdx + 4));
      else add(piece);
    }
  }

  return out;
}

function extractPython(text: string): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  for (const m of text.matchAll(PY_DECL_RE)) {
    const name = m[1]!;
    if (seen.has(name)) continue;
    seen.add(name);
    out.push(name);
  }
  return out;
}

function isDir(path: string): boolean {
  try {
    return statSync(path).isDirectory();
  } catch {
    return false;
  }
}

function cmp(a: string, b: string): number {
  return a < b ? -1 : a > b ? 1 : 0;
}

function safeRead(path: string): string | null {
  try {
    return readFileSync(path, 'utf-8');
  } catch {
    return null;
  }
}
